src/vs/workbench/contrib/notebook/browser/contrib/multicursor/notebookMulticursor.ts

```diff
@@ -623,9 +630,219 @@ export class NotebookMultiCursorController extends Disposable implements INotebo
+				match.cell.setSelections(newSelections);
+			}
+
+			const textModel = await match.cell.resolveTextModel();
+			textModel.pushStackElement();
+
+			const editorConfig = this.constructCellEditorOptions(match.cell);
+			const rawEditorOptions = editorConfig.getRawOptions();
+			const cursorConfig: NotebookCursorConfig = {
+				cursorStyle: cursorStyleFromString(rawEditorOptions.cursorStyle!),
+				cursorBlinking: cursorBlinkingStyleFromString(rawEditorOptions.cursorBlinking!),
+				cursorSmoothCaretAnimation: rawEditorOptions.cursorSmoothCaretAnimation!
+			};
+
+			this.trackedCells.push({
+				cellViewModel: match.cell,
+				initialSelection: initialSelection,
+				matchSelections: newSelections,
+				editorConfig: editorConfig,
+				cursorConfig: cursorConfig,
+				decorationIds: [],
+				undoRedoHistory: this.undoRedoService.getElements(match.cell.uri)
+			});
+		}
+
+		this._nbIsMultiSelectSession.set(true);
+		this.state = NotebookMultiCursorState.Selecting;
+		this._nbMultiSelectState.set(NotebookMultiCursorState.Selecting);
+
+		// manually handle the new anchor cell process to avoid race conditions
+		await this.updateCursorsControllers();
+		this.updateAnchorListeners();
+	}
+
+	private async handleCellEditorSelectAllMatches(notebookTextModel: NotebookTextModel, cell: ICellViewModel) {
+		// can be triggered mid multiselect session, or from idle state
+		if (this.state === NotebookMultiCursorState.Idle) {
+			// get word from current selection + rest of notebook objects
+			const textModel = cell.textModel;
+			if (!textModel) {
+				return;
+			}
+			const inputSelection = cell.getSelections()[0];
+			const word = this.getWord(inputSelection, textModel);
+			if (!word) {
+				return;
+			}
+			this.word = word.word;
+			const index = this.notebookEditor.getCellIndex(cell);
+			if (index === undefined) {
+				return;
+			}
+			this.startPosition = {
+				cellIndex: index,
+				position: new Position(inputSelection.startLineNumber, word.startColumn),
+			};
+
+			this.anchorCell = this.notebookEditor.activeCellAndCodeEditor;
+			if (!this.anchorCell || this.anchorCell[0].handle !== cell.handle) {
+				throw new Error('Active cell is not the same as the cell passed as context');
+			}
+			if (!(this.anchorCell[1] instanceof CodeEditorWidget)) {
+				throw new Error('Active cell is not an instance of CodeEditorWidget');
+			}
+
+			// get all matches in the notebook
+			const findResults = notebookTextModel.findAllMatches(this.word, false, true, USUAL_WORD_SEPARATORS);
+
+			// create the tracked matches for every result, needed for cursor controllers
+			this.trackedCells = [];
+			for (const res of findResults) {
+				const resultCellViewModel = this.notebookEditor.getCellByHandle(res.cell.handle);
+				if (!resultCellViewModel) {
+					continue;
+				}
+
+				const initialSelection = resultCellViewModel.getSelections()[0];
```

Comment:
this part of logic shows up 3 times (1 in `handleFindWidgetSelectAllMatches` and 2 in `handleCellEditorSelectAllMatches`), we can use some abstraction here.